### ISSUE TITLE
Default per-container ulimits are too generous

### ISSUE THREAD

**Issue by kolyshkin** [moby/moby#38814]
## TL;DR
```
$ ulimit -n -u
open files                      (-n) 1024
max user processes              (-u) 62435

$ docker run --rm ubuntu bash -c "ulimit -n -u"
open files                      (-n) 1048576
max user processes              (-u) unlimited
```

## Problem

ulimit, being an archaic resource management mechanism (see [this pdf](https://www.socallinuxexpo.org/scale7x/sites/scale7x.socallinuxexpo.org/files/openvz-scale7x-resource-management.pdf), slide 18), is not completely obsoleted by cgroup controllers, and it is still an essential part of system administration.

Default ulimits for a new container are derived from those of dockerd itself. They are set in `docker.service` systemf file to `unlimited` values:

```
$ grep ^Limit /lib/systemd/system/docker.service
LimitNOFILE=infinity
LimitNPROC=infinity
LimitCORE=infinity
```

This is required for dockerd itself, but is way too generous for containers it runs. For comparison, ulimits for a user (including root) on the host system are pretty modest (this is an example from Ubuntu 18.04):
```
$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 62435
max locked memory       (kbytes, -l) 16384
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 62435
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

This can create a number of problems, such as container abusing system resources (e.g. DoS attacks). In general, cgroup limits should be used to prevent those, yet I think ulimits should be set to a saner values.

In particular, `RLIMIT_NOFILE`, a number of open files limit, which is set to 2^20 (aka 1048576), causes a slowdown in a number of programs, as they use the upper limit value to iterate over all potentially opened file descriptors, closing those (or setting CLOEXEC bit) before every fork/exec. I am aware of the following cases:

* rpm, reported in https://github.com/moby/moby/issues/23137, https://bugzilla.redhat.com/show_bug.cgi?id=1537564, fix: https://github.com/rpm-software-management/rpm/pull/444 (fixed in Fedora 28).
* python2, reported in https://github.com/docker/for-linux/issues/502, proposed fix: https://github.com/python/cpython/pull/11584 (WONTFIX as python2 is frozen)
* python's pexpect/ptyprocess library, reported in https://github.com/pexpect/ptyprocess/issues/50.

Attacking those one by one proved complicated and not very fruitful, as some software is obsoleted, some is hard to fix, etc. In addition, the above list is not a concise one, so there might be more cases like this we're not aware of.

## Workaround

It is sufficient to add something like the following to dockerd configuration, and do `systemctl reload docker`:

```json
kir@kd:~/go/src/github.com/docker/docker$ cat /etc/docker/daemon.json-ulimits 
{
	"default-ulimits": {
		"nofile": {
			"Name": "nofile",
			"Hard": 1024,
			"Soft": 1024
		},
		"nproc": {
			"Name": "nproc",
			"Soft": 65536,
			"Hard": 65536
		}
	}
}
```

(side note: yes, the format is ugly but that's what we have) 

## Proposed solution

Add built-in (e.g. compiled-in) default ulimits for `RLIMIT_NOFILE` and `RLIMIT_NPROC`, so that, unless set from daemon.json or dockerd cli, those are being applied to all containers.

**Comment by kolyshkin** [comment]
Related (and might be an alternative approach to fix this): https://github.com/docker/for-linux/issues/73

**Comment by kolyshkin** [comment]
Well, my original description that containers' ulimits are those of (or derived from) dockerd was wrong -- it is  those of containerd, not dockerd.

**Comment by kolyshkin** [comment]
OK, it looks like higher RLIMIT_NOFILE (aka LimitNOFILE) is required for BOTH containerd and dockerd, otherwise the default limit of 1024 is hit pretty fast, leading to inability to start more than ~200 "busybox top" containers, with the following error:

docker: Error response from daemon: OCI runtime create failed: container_linux.go:336: starting container process caused "process_linux.go:402: container init caused \"process_linux.go:385: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: time=\\\\\\\"2019-02-28T18:23:51-08:00\\\\\\\" level=fatal msg=\\\\\\\"open /var/run/docker/netns/d19c91c8c8e9: too many open files\\\\\\\"\\\\n\\\"\"": unknown.

**Comment by kolyshkin** [comment]
Now, setting `LimitNPROC=1024` for containerd and NOT setting it for dockerd (leaving at default of `62435`) does not seem to affect its functionality in any way (except, of course, that `ulimit -n` shows `1024` inside a container).

**Comment by cpuguy83** [comment]
nproc is rough... What if we just started to enforce a (configurable) default pids-limit?

**Comment by cpuguy83** [comment]
I guess the pids-limit doesn't really fix things like the python issue...

**Comment by thaJeztah** [comment]
There's a note in the systemd unit file that setting a limit causes a performance regression due to accounting; is that still accurate (and does that apply to the daemon, or the container?)

**Comment by polarathene** [comment]
Just experienced this while trying to figure out why a small third-party package `postsrsd` was showing heavy CPU usage where it appeared to be doing nothing (_it normally binds to ports of an interface very quickly_) later discovered to eventually be successful after running for 10 minutes.

It iterated through over 1 billion `close()` calls. Using the container option `--ulimit 'nofile=1024'` fixed that.

Is it safe to say that using `--ulimit` with that value is usually fine, in the sense that you'd know if you need a larger limit? (_eg: software used documents such, or you're writing your own software that requires it_)

**Comment by corhere** [comment]
I think there's a defensible default for `RLIMIT_NOFILE` in containers: 1024 (2**10) as the soft limit, and the value of `/proc/sys/fs/nr_open` ([with a fallback of 1048576 (2**20)](https://www.kernel.org/doc/Documentation/sysctl/fs.txt#:~:text=nr_open%3A%0A%0AThis%20denotes%20the%20maximum%20number%20of%20file%2Dhandles%20a%20process%20can%0Aallocate.%20Default%20value%20is%201024*1024%20(1048576)%20which%20should%20be%0Aenough%20for%20most%20machines.%20Actual%20limit%20depends%20on%20RLIMIT_NOFILE%0Aresource%20limit.)) as the hard limit. The soft limit of 1024 would resolve issues with programs which loop through the entire file-descriptor range, and is the largest value compatible with software which uses `select(2)`. Maxing out the hard limit would permit programs which know what they're doing (e.g. systemd, databases, Go 1.19+) to raise their soft limit automatically, without needing any privileged capabilities or user tuning of the container rlimits. [This is what systemd already does](http://0pointer.net/blog/file-descriptor-limits.html) so we can be be confident that these defaults would be be widely compatible with any software which runs without issue on modern Linux distros.

On the other hand, picking a default `RLIMIT_NPROC` which will satisfy the majority of users could be tricky. It [appears to be dynamically calculated at boot](https://elixir.bootlin.com/linux/v6.2/source/kernel/fork.c#L933) based on [the amount of system memory.](https://elixir.bootlin.com/linux/v6.2/source/kernel/fork.c#L867) We might be able to do some dynamic introspection of the running host system such as [parsing `/proc/1/limits`](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=Since%20Linux%202.6.36%2C%20this%20file%20is%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20readable%20by%20all%20users%20on%20the%20system.) for the hard `RLIMIT_NPROC` and applying that as the default hard+soft limits for each container. I don't know what would be a sensible fallback if procfs is not mounted, however. Falling back to today's status quo of "inherited from containerd" would be surprising and hard to troubleshoot, so maybe the only sensible fallback is to refuse to start any containers unless there is an explicit `RLIMIT_NPROC` set on the container or in the daemon config?

Judging by the number of issues opened and known compatibility problems caused by `RLIMIT_NOFILE` being too generous compared to `RLIMIT_NPROC`, I think we could get away with tackling just the `RLIMIT_NOFILE` defaults and leaving `RLIMIT_NPROC` as-is for now.

**Comment by thaJeztah** [comment]
Some additional comments also posted in https://github.com/docker/for-linux/issues/73#issuecomment-1455386927

**Comment by polarathene** [comment]
> Some additional comments

I have another big comment almost ready :sweat_smile: I believe I've finally been able to connect all the dots between maintainers + bugs over the years :+1: 

I'll add a link here to it as well once it's ready :smile:

**Comment by thaJeztah** [comment]
Thanks @polarathene ! I haven't fully read all the links you provided in that comment, but thanks for digging into both history on "our" side, as well as changes since in the kernel and systemd! It's definitely possible that some of the reasons for setting these options no longer apply to more current distros (but we may have to verify the older "LTS" kernels, such as the ones used by RHEL/CentOS 7).

**Comment by polarathene** [comment]
I finally published it, apologies for the verbosity though: https://github.com/containerd/containerd/pull/7566#issuecomment-1461134171

It's mostly what everyone already knows about the topic I think, but consolidates that knowledge, along with:
- The history of changes for `LimitNOFILE`.
- The late 2018 release of systemd v240 and the changes that introduced.
- There's also the Aug 2022 Go 1.19 release, ~~but I'm not sure if the implicit behaviour it introduced to raise soft limit to hard limit would apply to `dockerd` / `containerd`~~. 
   **EDIT:** Seems to apply to both.
   - `containerd.service` with no `LimitNOFILE` was applying a soft and hard limit to containers `524288` (_systemd default hard limit_)
   - Whereas an explicit `LimitNOFILE=1024:1048576` shows that **soft limit is raised to the hard limit for containers**.
   - Also visible via `cat /proc/$(pidof containerd)/limits`.
- Some distros opt-out of the systemd v240 change (Debian based), sticking to the default `fs.nr_open == 1048576`. Others may override `LimitNOFILE` in `docker.service` (_or user applied via systemd drop-ins or `/etc/sysctl.conf`_).

---

**EDIT:** This section was before realizing a separate soft limit value presently has no effect. Presumably since Go 1.19 (Aug 2022).

`LimitNOFILE=1024:524288` would align with systemd v240 and onwards. It could potentially be dropped entirely if that were a min supported version?

`LimitNOFILE=1024:1048576` should remedy most issues caused by `infinity`.
- Keeps the hard limit as it once was before, but ensuring a soft limit as if run without a container
- The separate soft limit should reduce the need to adjust via `--ulimit` for reasons cited against `1048567`.

In production environments:
- There is potential for encountering errors from reaching the hard limit, but that could be raised with a `LimitNOFILE` config override.
- Errors are an easier troubleshooting indicator to resolve, and production environments generally better equipped to adjust to business needs.

---

> It's definitely possible that some of the reasons for setting these options no longer apply to more current distros (but we may have to verify the older "LTS" kernels, such as the ones used by RHEL/CentOS 7).

At a quick glance, are you referring to 3.10 kernel series?
- You'd also want to consider systemd version they can support, which [this Sep 2022 blog post](https://www.sobyte.net/post/2022-09/systemd-update/) notes as not being able to reach v240 and the newer cgroup v2 support present since kernel 4.5.
- I am not familiar with policy for EOL on that support, but you'd need to drop it at some point?
  If necessary document additional steps for those distros - or if not installed through a third-party source, have the package maintainers for those distro releases adjust the `LimitNOFILE` accordingly?

**Comment by polarathene** [comment]
**UPDATE:** Confirmed that Go 1.19 appears to be raising the soft limit to hard implicitly, unless this was intentionally added to `docker`/`containerd` since my old VM versions?

**Current system:** Docker v23, containerd `1.6.19` on host ArchLinux (Kernel 6.2 + systemd v253):
- `containerd.service` with no `LimitNOFILE` was applying a soft and hard limit to containers `524288` (_systemd default hard limit_)
- Whereas an explicit `LimitNOFILE=1024:1048576` shows that **soft limit is raised to the hard limit for containers**.
- Also visible via `cat /proc/$(pidof containerd)/limits`.

**Old VM:** Docker `18.09` and containerd `1.2.5` on host Ubuntu 18.04 (kernel 4.15 + systemd 237):
- Both had `LimitNOFILE=infinity` set in systemd service files resolving to `fs.nr_open` of `1048576`.
- Set `LimitNOFILE=2048:262144` to the `containerd.service` + `systemctl daemon-reload && systemctl restart containerd` and ran `cat /proc/$(pidof containerd)/limits` + `docker run --rm -it alpine ash -c 'ulimit -Sn; ulimit -Hn'`; Both matched the expected `soft:hard` limits specified.
- Confirms that **since Go 1.19, a regression has been introduced for `LimitNOFILE` usage**, preventing configuring the soft limit here. Not a regression to shipped default `.service` configs where a single limit value was assigned, but could affect production environments that know better? (_assuming a `1024` soft limit on `containerd` is ok, if `containerd` needs more it should raise for itself and lower again for containers, perhaps as a default to `--ulimit`?_)
- For reference, if `LimitNOFILE` is omitted from `containerd.service` on this system, soft/hard limit falls back to `1024:4096` (2011 kernel default). RAM is 2GB, `fs.file-max` only `197291`.
- Fixing the implicit raising behaviour would allow for omitting `LimitNOFILE` in `containerd.service` from systemd v240 onwards, it would inherit `1024:524288` (_Debian too despite it's `infinity` difference_).

**Debian 11 (Bullseye):** Docker v23, containerd `1.16.18` with Kernel 5.10 + systemd 247:
- Same experience as with "Current system" above, with the exception of `infinity` resolving to `2^20` (`1048576`, value of `fs.nr_open`) instead of `2^30` (`1073741816`) which is intentional by Debian (_opt-out from systemd raising `fs.nr_open` to avoid `infinity` being resolved to excessive number_).
- Still inherits the systemd system service default of `1024:524288` if `LimitNOFILE` is removed from the `.service`. But `containerd` built with Go 1.19 presently raises soft limit to `2^19` (`524288`) as well which is inherited by containers.
- Host resolves soft and hard limit of `1024:1048576` (`ulimit -Sn; ulimit -Hn`).

**Comment by paralin** [comment]
FYI this issue also breaks cups.

**Comment by cpuguy83** [comment]
Go is being updated to reset the soft limit to the original value when execing a new process.
Backports to 1.19 and 1.20 are being worked on. Unfortuantely this also requires bumping golang.org/x/sys after updating go.

**Comment by polarathene** [comment]
Below is a document I was working on but got distracted elsewhere. AFAIK it's fairly complete.

I will add that `docker.service` is relevant to `docker build`, which does support the `--ulimit` option as well, but when building with `docker bake` you need to rely on daemon config or `.service` to avoid issues. `dnf` on Fedora image for example would take considerably longer to install packages in a `RUN` instruction.

---

## Investigated limits impact and costs

I've played around with limits of `.service` and sharing observations here. Not all of this information was easy to come by and I spent a bit of time looking into it, so hopefully it's useful to someone :+1:

### TL;DR

`2^16` (65k) `busybox` containers estimated resource usage:
- 688k tasks + 206 GB (192 GiB) memory in `containerd` (_10.5 tasks + 3MiB per container_)
- Requiring at minimum `LimitNOFILE=262144` (`containerd.service`) + `LimitNOFILE=393216` (`docker.service`) - based on `4:1` + `6:1` service FDs needed per container ratio.
- 2.49 million open files (_`fs.file-nr` must be below `fs.file-max` limit_) - approx 38 FDs per container.
- 25 GiB memory for the containers cgroup (_approx 400KiB per container_).

`LimitNOFILE=524288` (_systemd default since v240_) should be more than enough for most systems as a sane default. This should be more than enough for both `docker.service` and `containerd.service` resource needs, capable of supporting 65k containers.

Containers that do need higher limits can explicitly declare that (_via `--ulimit` or equivalent_), as the upper bound is not impacted by `containerd.service`. The same can be done for lowering limits if necessary, both should rarely be necessary for most containers.

While `docker.service` and `containerd.service` need the higher soft limit (_enforced implicitly since Go 1.19_), it would be unlikely required for containers. An upcoming release of Go (with backports to 1.19) will implicitly restore the soft limit to `fork` / `exec` processes AFAIK. Until then, the Docker daemon can be configured with `default-ulimit` setting to enforce a `1024` soft limit on containers.

### System details

```
Fedora 37 VM 6.1.9 kernel x86_64 (16 GB memory)
Docker v23, containerd 1.6.18, systemd v251

# Additionally verified with builds before Go 1.19 to test soft limit lower than the hard limit:
dnf install docker-ce-3:20.10.23 docker-ce-cli-1:20.10.23 containerd.io-1.6.8
```

## Observations in `.service` files for `LimitNOFILE` 

On a fresh install (_via VM on Vultr_) there was approx 1800 file descriptors open (`sysctl fs.file-nr`). I used a shell loop to run `busybox` containers until failure and adjusted the `LimitNOFILE` for `docker.service` and `containerd.service` to collect metrics for insights.

I noticed a consistent ratio of number of FDs needed per container:
- `docker.service` - `6:1` ratio (_`5:1` with `--network=host`_), approx 853 containers with `LimitNOFILE=5120` (_1024 with host network_).
- `containerd.service` - `4:1` ratio (_I did not verify if `--network=host` reduced this_), `LimitNOFILE=1024` should be capable of 256 containers, provided `docker.service` is also high enough (eg: `LimitNOFILE=2048`_).

In `containerd.service` there was also a clear pattern in resources per container, where the `LimitNOFILE` value, image used (_`busybox`, `alpine`, `debian`_), and number of containers remained constant:
- Each containers systemd `.scope` has 1 task and approx 400KiB memory (little bit less for `alpine` and `debian`).
- 10.5 tasks + 3MiB memory added per container to `systemctl status containerd` report.
- Approx 38 open files per container running (_`fs.file-nr` after, minus the before value, divided by number of containers_).
- `mailserver/docker-mailserver:edge` was also tested to compare to the `sleep 180` containers:
  - 33 tasks per container `.scope` and 85MiB memory reported via `systemd-cgtop` (_10GiB needed min to run 120 of these containers_).
  - In `containerd` per container average resources were 11 tasks + 3.4MiB memory (_approx 400MiB usage for 120 of these containers_). Roughly consistent with the lighter images resource usage in `containerd`.
  - Files opened per container also increased to 291 (_approx 35k files open for 120 of these containers_).
  - If you want to reproduce for this image, `docker run` should include these extra options: `--hostname example.test --env SMTP_ONLY=1` (_hostname required to init, `SMTP_ONLY=1` skips needing an account configured_).

Operations like `docker stats` need to open as many file descriptors as total containers running, otherwise it'll hang waiting. You can observe if the daemon has reached the limit with `ls -1 /proc/$(pidof dockerd)/fd | wc -l`.

## Reproduction

Set `LimitNOFILE=768` in `docker.service`, then `systemctl daemon-reload && systemctl restart docker`. You can confirm the limit is applied to the daemon process with `cat /proc/$(pidof dockerd)/limits`.

Running the following should list:
- How many containers are running.
- Number of open files.
- How many tasks and memory both the `containerd` and `dockerd` daemons are using.

```bash
# Useful to run before the loop to compare against output after the loop is done
(pgrep containerd-shim | wc -l) && sysctl fs.file-nr \
  && (echo 'Containerd service:' && systemctl status containerd | grep -E 'Tasks|Memory') \
  && (echo 'Docker service:' && systemctl status docker | grep -E 'Tasks|Memory')
```

Running this loop below should fail on the last few containers, about 123 should be created:

```bash
# When `docker.service` limit is the bottleneck, you may need to `CTRL + C` to exit the loop
# if it stalls while waiting for new FDs once exhausted and outputting errors:
for i in $(seq 1 130); do docker run --rm -d busybox sleep 180; done
```

You can add additional options:
- `--network host`:
  - Avoids creating a new veth interface (see `ip link`) to the default Docker bridge each `docker run`.
  - Without this `docker run` may fail after `1023` interfaces are present on the bridge?
  - Creation will be a bit faster, and the FD to container ratio for `dockerd` is lowered to `5:1`.
- `--ulimit "nofile=1023456789"`:
  - Useful to observe that it does not affect memory usage on it's own.
  - Also shows `dockerd` + `containerd` limits don't affect how high this can go.
  - For Debian based distros this would fail as it's higher than `fs.nr_open` (`1 048 576`), use that or a lower value.
- `--cgroup-parent=LimitTests.slice`:
  - Similar to `docker stats` but isolated from other running containers. `systemd-cgtop` does include disk cache (_file-backed memory pages_) in it's reported memory usage however (_use `sync && sysctl vm.drop_caches=3` to clear that_).
  - This can be useful if you want a better overview of resource usage across all the containers created: 
    - Create a temporary slice for testing with: `mkdir /sys/fs/cgroup/LimitTests.slice`.
    - Run `systemd-cgtop --order=memory LimitTests.slice` to only view the containers running in this cgroup slice sorted by memory usage.
    - Memory usage shown for the entire slice and per container. A `busybox` container uses roughly 400KB per container.


### Limits impact across process children

I had a misconception that child processes contributed to the parents open files limit.  However as my notes in this section detail, it's only inheriting the limits applied, each process seems to have it's own individual count.

Although, I'm probably missing something here as I have read of processes passing down FDs to children, which is also why daemons have a common hygiene practice to close the FD range available I think? This is lower level than I'm familiar with :sweat_smile:

- You can also observe the number of file descriptors open for the `dockerd` and `containerd` processes like this: `ls -1 /proc/$(pidof dockerd)/fd | wc -l`.
- This isn't applicable to the `containerd-shim` process that is responsible for the container, so `ls -1 /proc/$(pgrep --newest --exact containerd-shim)/fd | wc -l` won't be useful there.

To confirm this, run a container to test with: `docker run --rm -it --ulimit "nofile=1024:1048576" alpine ash`. Then try the following:

```bash
# Create a folder to add many files to:
mkdir /tmp/test; cd /tmp/test

# Create empty files:
for x in $(seq 3 2048); do touch "${x}.tmp"; done

# Open files to specific file descriptors:
for x in $(seq 1000 1030); do echo "${x}"; eval "exec ${x}< ${x}.tmp"; done
# Fails after 1024 because soft limit is capped there. Raise it:
ulimit -Sn 2048

# Now the loop before will be successful.
# You could cover the whole original soft limit range (excluding FDs 0-2: stdin, stdout, stderr):
for x in $(seq 3 1024); do echo "${x}"; eval "exec ${x}< ${x}.tmp"; done

# Multiple container processes / children opening as many files:
# You can run the same loop in a new shell process with `ash -c 'for ... done'`
# Or `docker exec` into the container from another terminal and run the loop at `/tmp/test` again.
# Each can open files up to their current soft limit, it doesn't matter what limits are set on `dockerd`, `containerd` or the containers PID 1.

############
### Tips ###
############

# You can observe the current limit applied:
cat /proc/self/limits
# And if you have not exhausted your FDs soft limit (due to the pipe),
# this will report how much of the limit is used:
ls -1 /proc/self/fd | wc -l
# Otherwise, outside of the container if this is your only `ash` process running,
# you can query it's PID to get this information:
ls -1 /proc/$(pgrep --newest --exact ash)/fd | wc -l

# Process count in container:
# `docker stats` should list a containers PIDs count for number of processes,
# `systemd-cgtop` should report the same value in it's Tasks column.
# Alternatively if you know the cgroup name like `docker-<CONTAINER_ID>.scope`:
# (NOTE: Path may differ if you used `--cgroup-parent`)
cat /sys/fs/cgroup/system.slice/docker-<CONTAINER_ID>.scope/pids.current

# List the processes with their PIDs:
# For a single container, you can visualize the process tree:
pstree --arguments --show-pids $(pgrep --newest --exact containerd-shim)
# Alternatively if you know the cgroup name like `docker-<CONTAINER_ID>.scope`:
systemd-cgls --unit docker-<CONTAINER_ID>.scope

# You can also observe disk cache in memory monitoring by creating a 1GB file:
dd if=/dev/zero of=bigfile bs=1M count=1000
free -h
# `systemd-cgtop` will include 1GB more in memory usage for the container,
# while `docker stats` appears to account only 30MiB (scales proportionally).
# Now clear the cache outside of the container and observe memory usage again:
sync && sysctl vm.drop_caches=3
```

You will notice that:
- Each process adds those FDs to the open file count returned from `fs.file-nr`, and frees them when that process is closed.
- You can re-run the loops for the same process and observe no change, the files are already counted as open for that process.
- There is a **memory cost involved**:
  - Each file `touch` costs about `2048` bytes (_disk-cache only until opened_).
  - Each file open (_1 or more FD references each increment `fs.file-nr`_) costs about `512` bytes per FD open for it.
  - Creating 512k files this way uses approx 1.1GiB memory (_not released with `sysctl vm.drop_caches=3` while opened by at least one FD_), while each process opening the equivalent amount of file descriptors additionally uses 250MiB (262MB).

## Errors

Nothing useful here, other than depending on which service limit was exhausted first resulted in slightly different errors.

Sometimes this made any `docker` command like `docker ps` hang (daemon exhausted limit). I had also observed:
- Containers not running (_no `pgrep containerd-shim` output, but `docker ps` listed containers running well beyond when they should have exited_).
- Containers running with `containerd-shim` process (using memory), despite `systemctl stop docker containerd`. Sometimes this needed `pkill containerd-shim` to cleanup and `systemctl start docker containerd` would log a bunch of errors in `journalctl` handling cleanup of dead shims (_depending on the number of containers, this may time out and need to start the `containerd` service again_).
- Even with all that out of the way, there was some memory usage of several hundred MB from the baseline that lingered. As it didn't seem to belong to any process, I assume it was kernel memory. I think the largest number of containers I experimented running with was around 1600-ish.

### `docker.service` limit exceeded

This failure output more errors per `docker run` but the errors varied:

```
ERRO[0000] Error waiting for container: container caff476371b6897ef35a95e26429f100d0d929120ff1abecc8a16aa674d692bf: driver "overlay2" failed to remove root filesystem: openfdat /var/lib/docker/overlay2/35f26ec862bb91d7c3214f76f8660938145bbb36eda114f67e711aad2be89578-init/diff/etc: too many open files 
```

```
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2023-03-12T02:26:20Z" level=fatal msg="failed to create a netlink handle: could not get current namespace while creating netlink socket: too many open files": unknown.
```

```
docker: Error response from daemon: failed to initialize logging driver: open /var/lib/docker/containers/b014a19f7eb89bb909dee158d21f35f001cfeb80c01e0078d6f20aac8151573f/b014a19f7eb89bb909dee158d21f35f001cfeb80c01e0078d6f20aac8151573f-json.log: too many open files.
```

### `containerd.service` limit exceeded

I think I've seen some others, but it's usually this one:

```
docker: Error response from daemon: failed to start shim: start failed: : pipe2: too many open files: unknown.
```

**Comment by antoinetran** [comment]
> ## Workaround
> 
> It is sufficient to add something like the following to dockerd configuration, and do `systemctl reload docker`:
> 
> ```json
> kir@kd:~/go/src/github.com/docker/docker$ cat /etc/docker/daemon.json-ulimits 
> {
> 	"default-ulimits": {
> 		"nofile": {
> 			"Name": "nofile",
> 			"Hard": 1024,
> 			"Soft": 1024
> 		},
> 		"nproc": {
> 			"Name": "nproc",
> 			"Soft": 65536,
> 			"Hard": 65536
> 		}
> 	}
> }
> ```

Hi, I would like to add 2 remarks related to that workaround.
1. Setting soft limit to 1024 is very low for dockerd. We encounteded an issue with that limit, because we have hundreds of containers in a host. One container takes 4 file descriptors (2 * stdin, 2 * stdout for json log). The open file limit in Docker 20 was not 1024 but 1048576. So I suggest to set as default the old open file limit to 1048576.
2. The workaround applied to /etc/docker/daemon.json only works in case Docker BuildKit is not enabled. When BuildKit is enabled, the ulimit is inherited only from Systemd. So the full workround is the modification of daemon.json and docker systemd
```
sudo mkdir -p /etc/systemd/system/docker.service.d
cat <<EOF | sudo tee /etc/systemd/system/docker.service.d/ulimit.conf
[Service]
# This ulimit is only applied when Docker BuildKit is used.
# See https://github.com/moby/buildkit/issues/1665
# For case when Docker BuildKit is not used, see /etc/docker/daemon.json instead.
LimitNOFILE=1048576:1048576
EOF
```
I hope these information should avoid other projects some hours of debugging!

**Comment by neersighted** [comment]
This should not be necessary with upcoming releases and a recent Go compiler -- Go will modify the limits for dockerd/containerd themselves, while preserving them for child processes, and systemd will set the system-level defaults to something sane for containers.

Also, limits on dockerd will not be inherited by containers except those run by BuildKit. Make sure to adjust limits on containerd as well.

For users of older versions, it is very possible that you may have to manually manage these for some time, however.

**Comment by polarathene** [comment]
> 1. We encounteded an issue with that limit, because we have hundreds of containers in a host.

You don't need to raise the limit so high, even without the improvements from Go 1.19+:
- A few hundred containers that you describe should be easily supported with `8192`.
- I [shared above earlier](https://github.com/moby/moby/issues/38814#issuecomment-1483951375) how `524288` (_systemd default for hard limit in `LimitNOFILE`_) will support 65k `busybox` containers, along with resource usage.
- You should not need anywhere near that much in practice. Github Actions CI uses `65536` IIRC (`2^16`) as both soft and hard limit, that should still support thousands of containers (_which will hit other bottlenecks on the system first_).

### Docker Engine earlier than [20.10.24](https://docs.docker.com/engine/release-notes/20.10/#201024)

Release `20.10.24` and `23.0.2` are both from March 2023 with Go `1.19.7` (_bug fix, earlier `23.0.x` releases require a workaround_). 

Any release prior to 2023 does not have the Go 1.19+ improvement, using a soft limit above `1024` may be necessary (_the default on these releases is set by `LimitNOFILE=infinity`, which would be `2^20` or `2^30` depending on Docker host_).

If the soft limit being high negatively affects processes in a container (_excessive performance regression or memory usage_), you can use `daemon.json`, or an explicit `--ulimit` option on containers. Setting `1024` for the soft limit on a container should work well (_you'd still want a decent hard limit which is fairly safe to have a large number, `524288` is usually more than enough for 99% of software_).

### Docker Engine releases in 2023

Usually only `dockerd` / `containerd` processes require this to be raised, not the majority of processes running within the container.

As explained by neersighted, newer versions of these that are built with Go in 2023 will do this for you (_some in 2022Q4, but there was a slight bug that would require `daemon.json` to set `default-ulimits` as a workaround_), and keep the default soft limit for processes in the container.

The systemd service file with `LimitNOFILE` is still relevant for now. If it is set to `infinity` you may want to override that.

---

**Current Status:**
- `LimitNOFILE=infinity` is still the case until Docker v25 (_not yet released_), unless the team is backporting it to any releases built with Go 1.19+.
- `containerd` still has not approved the equivalent change to remove `LimitNOFILE` from their systemd service file. It will likewise need overriding (_drop-in file via `systemctl edit containerd.service`_) until that is approved and pushed out into a release (_not sure if that'll get backported either_).
